mk/protobuf.mk

```diff
@@ -0,0 +1,83 @@
+# Copyright 2014 The Cockroach Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied. See the License for the specific language governing
+# permissions and limitations under the License. See the AUTHORS file
+# for names of contributors.
+#
+# Author: Tamir Duberstein (tamird@gmail.com)
+
+REPO_ROOT      := .
+ORG_ROOT       := $(REPO_ROOT)/..
+GITHUB_ROOT    := $(ORG_ROOT)/..
+GOGOPROTO_ROOT := $(GITHUB_ROOT)/gogo/protobuf
+
+ENGINE_ROOT := $(REPO_ROOT)/storage/engine
+
+# Ensure we have an unambiguous GOPATH
+RELATIVE_GOPATH := $(GITHUB_ROOT)/../..
+#                                 ^  ^~ GOPATH
+#                                 |~ GOPATH/src
+
+GOPATH_BIN      := $(RELATIVE_GOPATH)/bin
+PROTOC          := $(GOPATH_BIN)/protoc
+PROTOC_PLUGIN   := $(GOPATH_BIN)/protoc-gen-gogoroach
+GOGO_PROTO      := $(GOGOPROTO_ROOT)/gogoproto/gogo.proto
+GOGO_PROTO_PATH := $(GOGOPROTO_ROOT):$(GOGOPROTO_ROOT)/protobuf
+
+GOTOOLS := .go_tools
+CPPTOOLS := .cpp_tools
+
+GO_PROTOS  := $(sort $(shell find . -name *.proto -type f))
+GO_SOURCES  := $(GO_PROTOS:%.proto=%.pb.go)
+
+CPP_PROTOS := $(filter %api.proto %data.proto %internal.proto %config.proto %errors.proto,$(GO_PROTOS))
+CPP_HEADERS := $(subst ./,$(ENGINE_ROOT)/,$(CPP_PROTOS:%.proto=%.pb.h))
+CPP_SOURCES := $(subst ./,$(ENGINE_ROOT)/,$(CPP_PROTOS:%.proto=%.pb.cc))
+
+$(PROTOC_PLUGIN): $(REPO_ROOT)/protoc-gen-gogoroach/main.go $(shell find $(GOGOPROTO_ROOT))
+	go install github.com/cockroachdb/cockroach/protoc-gen-gogoroach
+
+.PHONY:
+protos: goprotos cppprotos
+
+.PHONY:
+goprotos: $(GOTOOLS) $(GO_SOURCES)
+
+.PHONY:
+cppprotos: $(CPPTOOLS) $(CPP_HEADERS) $(CPP_SOURCES)
+
+$(GOTOOLS): $(PROTOC) $(GOGO_PROTO) $(PROTOC_PLUGIN)
+	rm -f $(GO_SOURCES)
+	touch $(GOTOOLS)
+
+$(CPPTOOLS): $(PROTOC) $(GOGO_PROTO)
+	rm -f $(CPP_HEADERS) $(CPP_SOURCES)
+	touch $(CPPTOOLS)
```

Comment:
You seem to be jumping through hoops to avoid regenerating the protos, but I'm curious as to why. Proto generation is fairly fast and relatively rare. If you're worried about the subsequent recompiles of go code then I would suggest generating into a temp directory and comparing the old vs new before replacing.